src/wazuh_modules/wm_keyrequest.c

```diff
@@ -0,0 +1,458 @@
+/*
+ * Wazuh Module for remote key requests
+ * Copyright (C) 2018 Wazuh Inc.
+ * November 25, 2018.
+ *
+ * This program is a free software; you can redistribute it
+ * and/or modify it under the terms of the GNU General Public
+ * License (version 2) as published by the FSF - Free Software
+ * Foundation.
+ */
+
+#ifndef WIN32
+
+#include "wmodules.h"
+#include <os_net/os_net.h>
+#include "shared.h"
+
+#define RELAUNCH_TIME 10
+#define minfo(format, ...) mtinfo(WM_KEY_REQUEST_LOGTAG, format, ##__VA_ARGS__)
+#define mwarn(format, ...) mtwarn(WM_KEY_REQUEST_LOGTAG, format, ##__VA_ARGS__)
+#define merror(format, ...) mterror(WM_KEY_REQUEST_LOGTAG, format, ##__VA_ARGS__)
+#define mdebug1(format, ...) mtdebug1(WM_KEY_REQUEST_LOGTAG, format, ##__VA_ARGS__)
+#define mdebug2(format, ...) mtdebug2(WM_KEY_REQUEST_LOGTAG, format, ##__VA_ARGS__)
+
+static void * wm_key_request_main(wm_krequest_t * data);   // Module main function. It won't return
+static void wm_key_request_destroy(wm_krequest_t * data);  // Destroy data
+cJSON *wm_key_request_dump(const wm_krequest_t * data);     // Read config
+
+void * w_request_thread(const wm_krequest_t *data);
+
+// Dispatch request. Write the output into the same input buffer.
+static int wm_key_request_dispatch(char * buffer,const wm_krequest_t * data);
+
+static int external_socket_connect(char *socket_path, int repsonse_timeout);
+static void launch_socket(char *exec_path);
+
+/* Decode rootcheck input queue */
+static w_queue_t * request_queue;
+
+static OSHash *request_hash = NULL;
+
+const char *exec_params[2] = { "id", "ip" };
+
+const wm_context WM_KEY_REQUEST_CONTEXT = {
+    KEY_WM_NAME,
+    (wm_routine)wm_key_request_main,
+    (wm_routine)wm_key_request_destroy,
+    (cJSON * (*)(const void *))wm_key_request_dump
+};
+
+typedef enum _request_type{
+    W_TYPE_ID,W_TYPE_IP
+} _request_type_t;
+
+// Module main function. It won't return
+void * wm_key_request_main(wm_krequest_t * data) {
+    int sock;
+    int recv;
+    unsigned int i;
+    char buffer[OS_MAXSTR + 1];
+    char * copy;
+
+    // If module is disabled, exit
+    if (data->enabled) {
+        minfo("Module started");
+    } else {
+        minfo("Module disabled. Exiting.");
+        pthread_exit(NULL);
+    }
+
+    /* Ignore SIGPIPE, it will be detected on recv */
+    signal(SIGPIPE, SIG_IGN);
+
+    /* Init the request hash table */
+    request_hash = OSHash_Create();
+
+    if (!request_hash) {
+        merror(LIST_ERROR);
+        pthread_exit(NULL);
+    }
+
+    /* Init the queue input */
+    request_queue = queue_init(data->queue_size);
+
+    if ((sock = StartMQ(WM_KEY_REQUEST_SOCK_PATH, READ)) < 0) {
+        merror_exit(QUEUE_ERROR, WM_KEY_REQUEST_SOCK_PATH, strerror(errno));
```

Comment:
An issue creating a socket is a serious problem, but it should not stop the rest of modules.